### DISCUSSION TITLE
OAuth2 flows with RabbitMQ

### DISCUSSION THREAD

**Discussion by mdrio** [ag2ai/faststream#1915] [Q&A] [answered]
Hi, I am new to FastStream, I am trying to figure out how to manage OAuth2 flows with faststream and rabbit. Basically, I would like to reimplement this [code](https://github.com/rabbitmq/rabbitmq-oauth2-tutorial/blob/main/pika-client/producer.py). I am not sure how to pass the credentials (i.e. the access token) to the broker and refresh when required. 
Any ideas? Thanks.

**Comment by Lancetnik** [comment, marked as answer]
@mdrio 

You can try something like that, with lower aio-pika API access

```python
import asyncio
from contextlib import asynccontextmanager
from faststream import FastStream
from faststream.rabbit import RabbitBroker, security

broker = RabbitBroker()

async def update_secret_loop():
    while ...:
        new_token = ...
        broker._connection.update_secret(new_token)

@asynccontextmanager
async def oath_lifespan():
    initial_token = ...
    await broker.connect(security=security.SASLPlaintext("", initial_token))

    task = asyncio.create_task(update_secret_loop())
    yield
    task.cancel()

app = FastStream(broker, lifespan=oath_lifespan)
```

Unfortunately, we have no public `update_secret` method yet

**Reply by mdrio** [reply]
Thanks @Lancetnik, I'll try your solution.
